src/__tests__/react-context.js

```diff
@@ -0,0 +1,59 @@
+import React from 'react'
+import {render, Simulate} from '../'
+
+const StateContext = React.createContext()
+
+const MyButton = () => (
+  <StateContext.Consumer>
+    {({handleToggle}) => (
+      <button onClick={handleToggle}>
+        Push Me
+      </button>
+    )}
+  </StateContext.Consumer>
+)
+
+const SecretMessage = () => (
+  <StateContext.Consumer>
+    {({toggleStatus}) => (
+      <h1>
+        {toggleStatus ? 'Secret Message' : 'Hidden'}
+      </h1>
+    )}
+  </StateContext.Consumer>
+)
+
+class Container extends React.Component {
+  state = {
+    toggleStatus: false,
+  }
+
+  handleToggle = () => {
+    this.setState(prevState => ({
+      toggleStatus: !prevState.toggleStatus,
+    }))
+  }
+
+  render() {
+    return (
+      <div>
+        <StateContext.Provider
+          value={{
+            toggleStatus: this.state.toggleStatus,
+            handleToggle: this.handleToggle,
```

Comment:
TL;DR: if you could rewrite this so that `handleToggle` is in `state` (you'll have to move it above the state initialization) then pass `this.state` into `value`, that'd be great! Thanks!



This is a pretty common mistake with the context API. It's actually not a problem in this specific scenario because it's simple, but I'd prefer to avoid exacerbating the common mistake by having an example here :)

Anyway, the problem here is that you generally shouldn't have a `value` assigned to a value that is initialized in the `render` method. Every time the `value` is changed, React will re-render all consumers, even if the object is the same.

To solve this problem, the suggested solution is to put everything into `state` (including event handlers). That is, if you want to expose everything that's in `state` to consumers. If you don't, then the convention is to set a property on state called `context` and pass `this.state.context`, which actually makes things more complicated when updating the context, but avoids unnecessary re-renders.